src/shims/unix/mem.rs

```diff
@@ -113,21 +111,19 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
     ) -> InterpResult<'tcx, Scalar<Provenance>> {
         let this = self.eval_context_mut();
 
-        let addr = this.read_target_usize(addr)?;
+        let addr = this.read_pointer(addr)?;
         let length = this.read_target_usize(length)?;
 
         // addr must be a multiple of the page size
         #[allow(clippy::arithmetic_side_effects)] // PAGE_SIZE is nonzero
-        if addr % this.machine.page_size != 0 {
+        if addr.addr().bytes() % this.machine.page_size != 0 {
             this.set_last_error(Scalar::from_i32(this.eval_libc_i32("EINVAL")))?;
             return Ok(Scalar::from_i32(-1));
         }
 
         let length = round_to_next_multiple_of(length, this.machine.page_size);
 
-        let ptr = Machine::ptr_from_addr_cast(this, addr)?;
-
-        let Ok(ptr) = ptr.into_pointer_or_addr() else {
+        let Ok(ptr) = addr.into_pointer_or_addr() else {
```

Comment:
`addr` is already a `Pointer`. Why is this still needed?

Instead of `Machine::ptr_get_alloc`, you should call `self.ptr_get_alloc_id`.

Reply:
Actually I think we don't even want to do that much. I think we want to report UB in all the cases where we threw an unsupported error, because LLVM's semantics do make these UB.

Reply:
You need to get the alloc_id though and that is the method to get it.

Reply:
Oh, `deallocate_ptr` already does that. Yeah that also works.